ArduSub/mode_poshold.cpp

```diff
@@ -103,12 +83,51 @@ void ModePoshold::run()
             attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(target_roll, target_pitch, target_yaw_rate);
             sub.last_pilot_heading = ahrs.yaw_sensor; // update heading to hold
 
-        } else { // call attitude controller holding absolute absolute bearing
+        } else { // call attitude controller holding absolute bearing
             attitude_control->input_euler_angle_roll_pitch_yaw(target_roll, target_pitch, sub.last_pilot_heading, true);
         }
     }
 
-    // Update z axis //
+    // update z axis
     control_depth();
+
+    // update xy axis
+    // call this after Sub::get_pilot_desired_climb_rate is called so that THR_DZ is reasonable
+    control_horizontal();
+}
+
+void ModePoshold::control_horizontal() {
+    float lateral_out = 0;
+    float forward_out = 0;
+
+    // get desired rates in the body frame
+    Vector2f body_rates_cm_s = {
+        sub.get_pilot_desired_horizontal_rate(channel_forward),
+        sub.get_pilot_desired_horizontal_rate(channel_lateral)
+    };
+
+    if (sub.position_ok()) {
+        if (!position_control->is_active_xy()) {
+            // the xy controller timed out, re-initialize
+            position_control->init_xy_controller_stopping_point();
+        }
+
+        // convert to the earth frame and set target rates
+        auto earth_rates_cm_s = ahrs.body_to_earth2D(body_rates_cm_s);
+        position_control->input_vel_accel_xy(earth_rates_cm_s, {0, 0});
+
+        // convert pos control roll and pitch angles back to lateral and forward efforts
+        sub.translate_pos_control_rp(lateral_out, forward_out);
+
+        // udpate the xy controller
+        position_control->update_xy_controller();
+    } else {
+        // allow the pilot to reposition manually
+        forward_out = body_rates_cm_s.x / (float)g.pilot_speed;
```

Comment:
might want to put a check in here that the pilot_speed is non-zero

Reply:
Done